sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala

```diff
@@ -1026,7 +1026,14 @@ private[hive] object HiveClientImpl extends Logging {
     } else {
       CharVarcharUtils.getRawTypeString(c.metadata).getOrElse(c.dataType.catalogString)
     }
-    new FieldSchema(c.name, typeString, c.getComment().orNull)
+    val name = if (lowerCase) {
+      // scalastyle:off caselocale
+      c.name.toLowerCase
+      // scalastyle:on caselocale
+    } else {
+      c.name
```

Comment:
when shall we not lower-case it?

Reply:
In the last commit (https://github.com/apache/spark/pull/36027/commits/d7730e512497a09cf9a8564f4f47dd3447e8f6c4) I lowercased the data column names (and bucketspec columns) but kept the partitioning columns. This is because Hive seems to be case preserving with partitioning columns too (despite partition paths (`.../p=v/...`) are lowercased) and at many places in Spark we use that to match partitioning columns from Hive to `CatalogTable.partitionColumnNames`.

But even with this commit, when we lowercase data columns only, we have test failures like `Create Hive table as select should output correct schema` (https://github.com/apache/spark/blob/master/sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala#L942-L964) as we expect the schema of the parquet file to match the create table statement.

I feel we should stay on the safe side and simply just allign the bucketspec in `toHivetable` to the columns as the original commit (https://github.com/apache/spark/pull/36027/commits/f68ac14d1ff53c6b3104da5fb952495a97d5a4d1) did.

Reply:
Ah I see. This is very tricky. Although HMS is not case-preserving, but we still need to pass the original column names to Hive, as this may affect the column names in the schema of physical parquet files.

Reply:
Looking at the related code again, I feel it's unnecessary to have this `hive table -> spark table -> hive table` roundtrip when we need to pass the raw table to call `HiveClient` APIs. Shall we expose a `getHiveTable` function in `HiveClient` and let `HiveExternalCatalog` call it? This avoids the `spark table -> hive table` conversion and we should be able to fix this bug.

Reply:
I'm not sure we can add a method that uses `HiveTable` (`import org.apache.hadoop.hive.ql.metadata.{Table => HiveTable}`) to `HiveClient` as it contradicts to the purpose of `HiveClient`: https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClient.scala#L29-L33

If you feel the original commit (https://github.com/apache/spark/commit/f68ac14d1ff53c6b3104da5fb952495a97d5a4d1) is hacky then we can still fix the issue by passing fully restored catalog table to `HiveClient.getPartitionsByFilter()` in `HiveExternalCatalog .listPartitionsByFilter()`. But there might be other cases when need to do the fix, like `HiveExternalCatalog.renameTable()` seems to suffer from the same issue.

Reply:
We can use `Object` to store the raw hive table, so that we don't need to expose the Hive classes.